File: Math/NumberSystems/ChangeBases.py
# coding=utf-8
""" CHANGE THE BASE OF A NUMBER
    ob -> origin base
    od -> destiny base
"""

chars = "0123456789ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"

def changeBase(number, ob,od):
    if ob == 10:
        return tob(number, od)
    return tob(to10(number,ob),od)

""" FROM ANY BASE TO BASE 10
    b -> base of the number n
    pos -> location of a sub-number in n
"""
def to10(n, b, pos =0):
   if n == 0: return 0
   return (n % 10)* (b  ** pos) + to10(n / 10, b, pos+1)

"""FROM TEN BASE TO ANOTHER BASE"""
def tob(n, b):
    if n == 0: return ""
    return   tob(n // b, b) + chars[n % b]

def main():
    print ( tob(7,2))
    print ( tob(252,16))
    print ( tob(234,15))
    print ( to10(1000,2))
    print ( changeBase(111,2,10))
main()

File: Math/NumberSystems/ChangeBases.cpp
#include<bits/stdc++.h>
#define endl '\n'
#define show(x) cout <<#x << " =" <<x <<endl;
using namespace std;
typedef long long ll;
string chars = "0123456789ABCDEFGHIJKLMNÑOPQRSTUVWXYZ";

ll to10(ll n , ll b, ll mul){
  if (n ==0 ) return 0;
  return (n % 10)*mul + to10(n / 10, b, mul*1LL*b);
}

string tob(ll n, ll b){
  if (n == 0) return "";
  return  tob(n / b, b) + chars[n % b];
}
/*
 * ob -> origin base
 * db -> destiny base
 */
string changeBase(ll num, ll ob, ll db){
  if (ob == 10) return tob(num, db);
  return tob(to10(num, ob, 1LL), db);
}
int main(){
  cout << changeBase(1000,2,10) <<endl;
}
